fstab-generator/repart: make sure repart runs safely in the initrd and can generate a root fs there

poettering opened the pull request:
There's one model of preparing OS images that I am particularly interested in: a minimal image is shipped that consists of one /usr/ file system (plus its verity) plus an ESP and nothing else. On first boot, we then add in a second set of /usr/ file system (plus its verity) and add in an encrypted root file system (ideally even enrolled to the TPM), spanning the rest of the disk. This method has many benefits, but requires some love to actually work: so far the root fs is assumed to exist already at the time the initrd runs, and that /usr/ is mounted only after the root fs is mounted.

This PR adds some love to this, so that we can mount /usr/ of the host OS before the root fs (so that we can look into it to find repart files), then can run repart with that mounted, and then can mount the root fs we just created, bind mount the /usr we already have into it and then finally boot that up.

The steps to get there are:

1. First, if mount.usr= is specified on kernel cmdline, mount this to /sysusr/usr first, and then in a second mount bind mount that to /sysroot/usr. This allows us to mount the /usr fs independently of the root fs, so that we can look into it for repart data.
2. For this we need to tweak fsck handling, since the fsck for the /usr fs also needs to be able to run independently of the fsck for the rootfs (previously all fscsk for non-root-fs were ordered after rootfs' fsck)
3. A new generic target "initrd-usr-fs.target" is introduced, similar to "initrd-root-fs.target", that tells us when /sysusr/usr has completed mounting. We then use this to order systemd-repart generically after that.

There are some other fixups, and cleanups here and there. Mostly minor stuff and refactoring to make things prettier.

Why use /sysusr/usr as intermediary mount point for the host /usr/ fs instead of just /sysusr/? Mostly so that we can invoke various tools with their regular --root= switches to find files and directories in this tree. If we'd mount the /usr fs directly to /sysusr/ then they'd all have to learn special modes to operate relative to these /usr/-only mounts.

bluca commented on src/partition/repart.c:
```diff
@@ -3913,7 +3914,52 @@ static int acquire_root_devno(const char *p, int mode, char **ret, int *ret_fd)
+static int find_os_prefix(char **ret) {
+        char *p;
+        int r;
+
```
assert(ret) ? it's deref later

bluca commented:
CentOS fails look related?

```
initrd-usr-fs.target: Failed to load configuration: No such file or directory
```

```
Failed to process message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=SwitchRoot cookie=1 reply_cookie=0 signature=ss error-name=n/a error-message=n/a: Could not resolve init executable /usr/lib/systemd/systemd: No such file or director
```

```
minitrd-switch-root.service: Job 39 initrd-switch-root.service/start finished, result=failed
```

bluca reviewed (approved):
looks good to me in general, but one minor nit and most importantly CI seems unhappy

keszybz commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -767,8 +768,22 @@ static int add_sysroot_usr_mount(void) {
+        if (streq(arg_usr_what, "gpt-auto")) {
+                /* This is handled by the gpt-auto generator */
+                log_debug("Skipping /usr/ directory handling, as gpt-auto was requested.");
+                return 1;
```
Maybe add a comment what 0 and 1 means in the return value.

keszybz commented on src/fstab-generator/fstab-generator.c:
```diff
@@ -721,7 +721,7 @@ static int add_sysroot_mount(void) {
                 opts = arg_root_options;
 
-        log_debug("Found entry what=%s where=/sysroot type=%s", what, strna(arg_root_fstype));
+        log_debug("Found entry what=%s where=/sysroot type=%s opts=%s", what, strna(arg_root_fstype), strna(opts));
```
I think this should be `strempty(opts)`. After all, it's fine for opts to be empty.

keszybz commented on src/partition/repart.c:
```diff
@@ -3913,7 +3914,52 @@ static int acquire_root_devno(const char *p, int mode, char **ret, int *ret_fd)
+        char *p;
+        int r;
+
+        /* Searches the right place to look for the OS root. This is relevant in the initrd: in the initrd
```
Searches **for**

keszybz commented on src/partition/repart.c:
```diff
@@ -3913,7 +3914,52 @@ static int acquire_root_devno(const char *p, int mode, char **ret, int *ret_fd)
+        else if (r > 0) {
+                log_debug("/sysroot/ is a mount point, assuming it's the prefix.");
+
+                p = strdup("/sysroot");
```
Since a static string or NULL is always returned, please just return `const char*`.

keszybz commented on src/partition/repart.c:
```diff
@@ -3913,7 +3914,52 @@ static int acquire_root_devno(const char *p, int mode, char **ret, int *ret_fd)
+                return 0;
+        }
+
+        r = path_is_mount_point("/sysusr/usr", NULL, 0);
+        if (r < 0 && r != -ENOENT)
+                log_debug_errno(r, "Failed to determine whether /sysusr/usr is a mount point, assuming it is not: %m");
+        else if (r > 0) {
+                log_debug("/sysusr/usr/ is a mount point, assuming /sysusr/ is the prefix.");
+
+                p = strdup("/sysusr");
+                if (!p)
+                        return -ENOMEM;
+
+                *ret = p;
+                return 0;
+        }
```
Hmm, shouldn't the order be reversed? Let's say we have already mounted `/sysroot`, but haven't gotten around to `/sysroot/usr` yet. We'll return a prefix where there is no data, even though it might be available under `/sysusr/usr`.

Also, returning -ENOENT seems risky... Why not just return `/sysroot`, to avoid a failure mode? It's not nice for the tool to fail if stuff is not mounted fully yet.

poettering replied:
true

keszybz commented on src/partition/repart.c:
```diff
@@ -4185,8 +4185,22 @@ static int run(int argc, char *argv[]) {
         if (in_initrd()) {
-                /* Default to operation on /sysroot when invoked in the initrd! */
-                arg_root = strdup("/sysroot");
+                /* Default to operation on /sysroot/ or /sysusr/ when invoked in the initrd. */
```
"to operating" ?

keszybz commented on src/shared/generator.c:
```diff
@@ -151,6 +151,13 @@ int generator_write_fsck_deps(
         assert(where);
 
+        /* Let's do an early exit if we are invoked for the root and /usr trees in the initrd, to not
+         * generated confusing log messages */
```
generated → generate, or "to avoid generating"

keszybz commented on src/shared/generator.c:
```diff
@@ -177,6 +177,11 @@ int generator_write_fsck_deps(
+                /* We support running the fsck instance for the root fs while it is already mounted, for
+                 * compatibility with non-initrd boots. It's ugly, but it is how it is. Since – unlike for
+                 * regular file systems – this means the ordering is reversed (i.e. mount *before* fsck) we
+                 * have a separate fsck unit for this, independent of systemd-fsck@.service*/
```
Missing ". " at the end.

poettering commented:
Rebased and fixed all raised issues.

I split out two PRs from this one: #19371 and #19372

The CI failure the earlier version of this PR saw is probably relevant, but deeply puzzling to me, hence by splitting these out and getting them merged ahead of time we can maybe track down what commit the CI failure was specifically caused by.

poettering commented:
Everything in this PR has now been split out. Open remain #19380 #19381 #19386

poettering commented:
Closing this one. There's one commit left, which exists in two variants in #19401 and #19380. No need to keep this one as a thrid.
